suprb/optimizer/solution/saga2/base.py

```diff
@@ -0,0 +1,170 @@
+        self.fitness_max = np.max([i.fitness_ for i in self.population_])
+
+        for iter_ in range(self.n_iter):
+
+            # Elitism
+            elitists = sorted(self.population_, key=lambda i: i.fitness_, reverse=True)[:self.n_elitists_]
+
+            # Selection
+            parents = self.selection(population=self.population_, n=self.population_size,
+                                     random_state=self.random_state_)
+            
+            # Note that this expression swallows the last element, if `population_size` is odd
+            parent_pairs = map(lambda *x: x, *([iter(parents)] * 2))
+
+            # Crossover
+            children = list(flatten([(self.crossover(A, B, self.crossover_rate_current_min, self.crossover_rate_max, self.fitness_mean, self.fitness_min, self.fitness_max, random_state=self.random_state_),
+                                      self.crossover(B, A, self.crossover_rate_current_min, self.crossover_rate_max, self.fitness_mean, self.fitness_min, self.fitness_max, random_state=self.random_state_))
+                                     for A, B in parent_pairs]))
+            # If `population_size` is odd, we add the solution not selected for reproduction directly
+            if self.population_size % 2 != 0:
+                children.append(parents[-1])
+
+            # Mutation
+            for child in children: child.fit(X, y)
+            mutated_children = [self.mutation(child, self.mutation_rate_min, self.mutation_rate_current_max, self.fitness_mean, self.fitness_max, random_state=self.random_state_) for child in children]
+
+            # Keep copy of old generation for comparison
+            self.old_gen = self.population_
+
+            # Replacement
+            self.population_ = elitists
+            self.population_.extend(mutated_children)
+
+            self.fit_population(X, y)
+            self.fitness_mean = np.mean(([i.fitness_ for i in self.population_]))
+            self.fitness_min = np.min(([i.fitness_ for i in self.population_]))
+            self.fitness_max = np.max(([i.fitness_ for i in self.population_]))
+
+            # Adjust Rates for current fitness
+            if iter_ > 0 : self.adjust_rates()
+
+
+    def calc_similarity(self):
+        try:
+            dot_sum = np.sum([np.multiply(self.old_gen[i].fitness_, self.population_[i].fitness_) for i in range(len(self.population_))])
+            length_new = np.sqrt(np.sum([np.square(i.fitness_) for i in self.population_]) + np.exp(-10))
+            length_old = np.sqrt(np.sum([np.square(i.fitness_) for i in self.old_gen]) + np.exp(-10))
+            cosine_similarity = dot_sum / (length_new + length_old)
+            genome_similarity = np.intersect1d([i.fitness_ for i in self.population_], [i.fitness_ for i in self.old_gen]).size / np.union1d([i.fitness_ for i in self.population_], [i.fitness_ for i in self.old_gen]).size
+        except ZeroDivisionError:
+            return 1
```

Comment:
Why do we have this if you add `+ np.exp(-10)`? And why do we return `1` in that case? Shouldn't it be a huge value as well?